src/main/scala/viper/gobra/frontend/info/implementation/typing/ghost/GhostMiscTyping.scala

```diff
@@ -258,12 +258,39 @@ trait GhostMiscTyping extends BaseTyping { this: TypeInfoImpl =>
   }
 
   implicit lazy val wellDefSpec: WellDefinedness[PSpecification] = createWellDef {
-    case PFunctionSpec(pres, preserves, posts, _) =>
+    case n@ PFunctionSpec(pres, preserves, posts, terminationMeasures, _) =>
       pres.flatMap(assignableToSpec) ++ preserves.flatMap(assignableToSpec) ++ posts.flatMap(assignableToSpec) ++
-      preserves.flatMap(e => allChildren(e).flatMap(illegalPreconditionNode)) ++ 
-      pres.flatMap(e => allChildren(e).flatMap(illegalPreconditionNode))
+      preserves.flatMap(e => allChildren(e).flatMap(illegalPreconditionNode)) ++
+      pres.flatMap(e => allChildren(e).flatMap(illegalPreconditionNode)) ++
+      terminationMeasures.flatMap(wellDefTerminationMeasure) ++
+      // if has conditional clause, all clauses must be conditional
+      // can only have one non-conditional clause
+      error(n, "Specificaions can either contain one non-conditional termination measure or multiple conditional-termination measures.", terminationMeasures.length > 1 && !terminationMeasures.forall(isConditional)) ++
+      // measures must have the same type
+      error(n, "Termination measures must all have the same type.", !hasSameMeasureType(terminationMeasures))
+
+    case PLoopSpec(invariants, terminationMeasure) =>
+      invariants.flatMap(assignableToSpec) ++ terminationMeasure.toVector.flatMap(wellDefTerminationMeasure)
+  }
+
+  private def wellDefTerminationMeasure(measure: PTerminationMeasure): Messages = measure match {
+    case PTupleTerminationMeasure(tuple, cond) =>
+      tuple.flatMap(p => comparableType.errors(exprType(p))(p) ++ isWeaklyPureExpr(p)) ++
+        cond.toVector.flatMap(p => assignableToSpec(p) ++ isWeaklyPureExpr(p))
```

Comment:
is it possible to have a condition with resources? I think it should be strongly pure